### ISSUE TITLE
skipgrams should exclude true positive context word from being sampled as negative sample

### ISSUE THREAD

**Issue by goodcheer** [keras-team/keras#16395]
Please go to TF Forum for help and support:

https://discuss.tensorflow.org/tag/keras

If you open a GitHub issue, here is our policy:

It must be a bug, a feature request, or a significant problem with the documentation (for small docs fixes please send a PR instead).
The form below must be filled out.

**Here's why we have that policy:**.

Keras developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04.3 LTS
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.6.1
- Python version: 3.8.10
- Bazel version (if compiling from source): NA
- GPU model and memory: NA
- Exact command to reproduce: below

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with:
python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"

**Describe the problem**.

When doing negative sampling, the indices should be sampled from outside of current window, by definition.

**Describe the current behavior**.

However, in tf.keras.preprocessing.sequence.skipgrams, when sampling [center word index, context word index], context word index is sampled from whole range of index, including corresponding within-window context word indices. (line 353)

https://github.com/keras-team/keras/blob/4c87dc9685eea2ed20111f9604b10d627b17f032/keras/preprocessing/sequence.py#L347-L358

As a result, positive couples of [center word index,  within-window context word index] might have two opposing label (0: negative, 1: positive).

I was able to verify this issue with following simple code.
 
```
from tensorflow.keras.preprocessing.sequence import skipgrams

seq = [1, 2, 3, 4]

sgns = skipgrams(seq, 5, window_size=3, negative_samples=1)
sgps  = skipgrams(seq, 5, window_size=3, negative_samples=0)

def find_mislabeled(sg_arr):
    hmap = {}
    for couple, label in zip(*sg_arr):
        key = str(couple)
        if key in hmap:
            hmap[key].add(label)
        else:
            hmap[key] = {label,} 
    return {k: v for k, v in hmap.items() if len(v) > 1}
    

print(len(find_mislabeled(sgps)) == 0) # True
print(len(find_mislabeled(sgns)) == 0) # False

```

**Describe the expected behavior**.

exclude the context word from being sampled as negative sample.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no): yes
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**.
 
```
from tensorflow.keras.preprocessing.sequence import skipgrams

seq = [1, 2, 3, 4]

sgns = skipgrams(seq, 5, window_size=3, negative_samples=1)
sgps  = skipgrams(seq, 5, window_size=3, negative_samples=0)

def find_mislabeled(sg_arr):
    hmap = {}
    for couple, label in zip(*sg_arr):
        key = str(couple)
        if key in hmap:
            hmap[key].add(label)
        else:
            hmap[key] = {label,} 
    return {k: v for k, v in hmap.items() if len(v) > 1}
    

print(len(find_mislabeled(sgps)) == 0) # True
print(len(find_mislabeled(sgns)) == 0) # False

```

Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

**Source code / logs**.

Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Comment by sushreebarsa** [comment]
@gadagashwini I was able to reproduce the issue on colab using TF v2.8.0 , tf-nightly and please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/eed7b9c7b32442a1f0a28b8902831eed/16395-keras.ipynb).Thanks!

**Comment by goodcheer** [comment]
might be relevant to  #44758,

**Comment by LukeWood** [comment]
Hey @goodcheer!  If you are willing to submit a PR to fix this we can accept it.  Thanks for the bug report!

**Comment by tilakrayal** [comment]
@goodcheer,
I tried to execute the code on tf-nightly and the issue was still reproducible. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/f61f7e2670fccd7399f112bec1229881/untitled1132.ipynb) and as the developer suggested above, could you please try to submit the PR for the fix. Thank you!
